resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/engines/JettyClientEngine.java

```diff
@@ -0,0 +1,223 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLContext;
+import javax.ws.rs.ProcessingException;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.client.InvocationCallback;
+import javax.ws.rs.client.ResponseProcessingException;
+import javax.ws.rs.core.MultivaluedHashMap;
+import javax.ws.rs.core.MultivaluedMap;
+
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.api.Response;
+import org.eclipse.jetty.client.api.Result;
+import org.eclipse.jetty.client.util.DeferredContentProvider;
+import org.eclipse.jetty.http.HttpFields;
+import org.eclipse.jetty.util.Callback;
+import org.jboss.resteasy.client.jaxrs.AsyncClientHttpEngine;
+import org.jboss.resteasy.client.jaxrs.internal.ClientInvocation;
+import org.jboss.resteasy.client.jaxrs.internal.ClientResponse;
+
+public class JettyClientEngine implements AsyncClientHttpEngine {
+    private static final InvocationCallback<ClientResponse> NOP = new InvocationCallback<ClientResponse>() {
+        @Override
+        public void completed(ClientResponse response) {
+        }
+
+        @Override
+        public void failed(Throwable throwable) {
+        }
+    };
+
+    private final HttpClient client;
+
+    public JettyClientEngine(HttpClient client) {
+        if (!client.isStarted()) {
+            try {
+                client.start();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        }
+        this.client = client;
+    }
+
+    @Override
+    public SSLContext getSslContext() {
+        return client.getSslContextFactory().getSslContext();
+    }
+
+    @Override
+    public HostnameVerifier getHostnameVerifier() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ClientResponse invoke(ClientInvocation invocation) {
+        Future<ClientResponse> future = submit(invocation, true, NOP, null);
+        try {
+            return future.get(client.getIdleTimeout(), TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            future.cancel(true);
+            throw clientException(e, null);
+        } catch (ExecutionException | TimeoutException e) {
+            throw clientException(e.getCause(), null); // NOPMD
+        }
+    }
+
+    /**
+     * Implementation note: due to lack of asynchronous message decoders the request must either be buffered,
+     * or it must have a {@code null} extractor and type parameter {@code <T>} must be {@link ClientResponse},
+     * which will read the data through its stream.  It is not possible to use the synchronous JAX-RS message
+     * decoding infrastructure without buffering or spinning up auxiliary threads (arguably more expensive than buffering).
```

Comment:
The argument would depend on the size of the payload.

Reply:
Yeah, I agree -- we may end up adding the threaded option later, but it's not needed to ship the first version.